### PR TITLE
[Merged by Bors] - feat(NumberField/CanonicalEmbedding): define the `negAt` map

### PR DESCRIPTION
**Pull request by xroblot**
Let `s` be a set of real places of the number field `K`. This PR defines the map `negAt s` on `mixedSpace K` that swaps the sign at all real places in `s` and proves some of its properties. 

It will be used later in #18231. 

This PR is part of the proof of the Analytic Class Number Formula.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by Ruben-VandeVelde — APPROVED** [review]
maintainer merge

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 1]
```diff
@@ -746,4 +766,98 @@ theorem fundamentalDomain_idealLattice :
...
+
+/-- `negAt` preserves the `norm`. -/
+@[simp]
+theorem norm_negAt [NumberField K] (x : mixedSpace K) :
+    mixedEmbedding.norm (negAt s x) = mixedEmbedding.norm x :=
+  norm_eq_of_normAtPlace_eq (fun w ↦ normAtPlace_negAt _ _ w)
+
+open ContinuousLinearEquiv in
+/-- `negAt` is its own inverse. -/
+theorem negAt_symm :
```
Comment (jcommelin) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:837]: I think this could also be a simp lemma.

**Review by jcommelin — COMMENTED** [review]
bors d+

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 2]
```diff
@@ -216,6 +217,25 @@ theorem _root_.NumberField.mixedEmbedding_injective [NumberField K] :
...
+
+open Classical in
+instance : IsAddHaarMeasure (volume : Measure (mixedSpace K)) :=
+  prod.instIsAddHaarMeasure volume volume
+
+open Classical in
+instance : NoAtoms (volume : Measure (mixedSpace K)) := by
+  obtain ⟨w⟩ := (inferInstance : Nonempty (InfinitePlace K))
+  by_cases hw : IsReal w
+  · exact @prod.instNoAtoms_fst _ _ _ _ volume volume _ (pi_noAtoms ⟨w, hw⟩)
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:234]: Does `infer_instance`, or maybe `inferInstanceAs`, work here? If not, maybe the `@` can be avoided using `(... := ...).`

File: Mathlib/Topology/Algebra/Module/Basic.lean [hunk 3]
```diff
@@ -2001,6 +2001,32 @@ def arrowCongrEquiv (e₁₂ : M₁ ≃SL[σ₁₂] M₂) (e₄₃ : M₄ ≃SL[
     ContinuousLinearMap.ext fun x => by
       simp only [ContinuousLinearMap.comp_apply, apply_symm_apply, coe_coe]
 
+/-- Combine a family of continuous linear equivalences into a continuous linear equivalence of
+pi-types. -/
+def piCongrRight {ι : Type*} {M : ι → Type*} [∀ i, TopologicalSpace (M i)]
```
Comment (riccardobrasca) [inline Mathlib/Topology/Algebra/Module/Basic.lean:2006]: Can you use a `variable` line here? Maybe in a new small section (I don't know what is already in scope).

**Review by riccardobrasca — APPROVED** [review]
LGTM, thanks!

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean (hunk 2 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:234]: I did it in two steps so that there is no need for `@`

**Comment by riccardobrasca** [comment]
There is a merge conflict but otherwise I think it is ready to go.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**